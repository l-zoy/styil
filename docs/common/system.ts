import { createSystem } from 'styls-react'

export const { styled, SystemProvider, useSystem, getCssValue } = createSystem({})
