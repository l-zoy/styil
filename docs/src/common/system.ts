import { createSystem } from 'styls-react'

export const { styled, SystemProvider, useSystem } = createSystem({})
