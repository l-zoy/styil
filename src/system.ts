import React from 'react'
import { createSelector } from './createSelector'
import { type SystemOptions, type Styled, type System } from './systemTypes'
import { type AnyObject } from './types'
import { parseRules } from './parse'
import { StyleSheet } from './sheet'

const globalCache = new Set<string>([])
const cacheKey = '__styls_cache__'

export function createSystem<Theme extends AnyObject = {}>(
  options: SystemOptions<Theme> = {}
): System<Theme> {
  const {
    theme: inputTheme = () => ({} as AnyObject),
    defaultMode = 'none',
    sheetOptions = {}
  } = options
  const { key, container, speedy, nonce } = sheetOptions

  if (globalThis.document && !globalCache.size) {
    const meta = document.getElementById(cacheKey) as HTMLMetaElement
    meta?.content.split(',').forEach((name) => {
      globalCache.add(name)
    })
  }

  const sheet = new StyleSheet({
    key: key ?? 'css',
    speedy: speedy === undefined ? process.env.NODE_ENV === 'production' : speedy,
    container: globalThis.document ? container ?? globalThis.document.head : null,
    nonce
  })

  const themeContent = React.createContext<{
    mode: string
    setMode: React.Dispatch<React.SetStateAction<string>>
    theme: Theme
  }>(
    // @ts-expect-error no value initially
    {}
  )

  const SystemProvider = (props: { children: React.ReactNode }) => {
    const [mode, setMode] = React.useState<string>(defaultMode)

    return React.createElement(
      themeContent.Provider,
      { value: { mode, setMode, theme: inputTheme(mode) } },
      props.children
    )
  }

  const useSystem = () => {
    return React.useContext(themeContent)
  }

  const styled: Styled<Theme> = (tag, styles, interpolation) => {
    let inputTag = tag
    const inputNamespace = (tag as { namespace: string }).namespace ?? ''

    if (typeof tag === 'object') {
      inputTag = tag.tag
    }

    let currentMode: string = defaultMode
    const modeIdentifier: Record<string, { targetClassName: string; namespaceJoiner: string }> = {}

    function createRule(
      mode: string,
      inputTargetInfo: { targetClassName: string; namespaceJoiner: string }
    ) {
      const identifier = modeIdentifier[mode]

      if (identifier) {
        const { namespaceJoiner, targetClassName } = identifier
        inputTargetInfo.namespaceJoiner = namespaceJoiner
        inputTargetInfo.targetClassName = targetClassName
        return
      }

      const theme = inputTheme(mode)
      const style = typeof styles === 'function' ? styles(theme, mode) : styles
      const variants =
        typeof interpolation === 'function' ? interpolation(theme, mode) : interpolation

      const selector = createSelector(style)
      let targetClassName = selector
      let namespaceJoiner = ''

      if (inputNamespace) {
        targetClassName = `${inputNamespace}-${selector}`
        namespaceJoiner = `${inputNamespace}-`
      }

      if (!globalCache.has(targetClassName)) {
        globalCache.add(targetClassName)
        sheet.insertStyle(parseRules(style, `.${targetClassName}`))
      }

      if (variants) {
        const variantsKeys = Object.keys(variants)
        let variantsIndex = variantsKeys.length

        while (variantsIndex--) {
          const variantsKey = variantsKeys[variantsIndex]
          const variantsValue = variants[variantsKey]

          const variantsChildKeys = Object.keys(variantsValue)
          let variantsChildIndex = variantsChildKeys.length

          while (variantsChildIndex--) {
            const key = variantsChildKeys[variantsChildIndex]
            const value = variantsValue[key]
            const variantsClassName = `.${targetClassName}.${namespaceJoiner}${variantsKey}-${key}`

            if (!globalCache.has(variantsClassName)) {
              globalCache.add(variantsClassName)
              sheet.insertStyle(parseRules(value, variantsClassName))
            }
          }
        }
      }

      inputTargetInfo.targetClassName = targetClassName
      inputTargetInfo.namespaceJoiner = namespaceJoiner

      modeIdentifier[mode] = {
        targetClassName,
        namespaceJoiner
      }
    }

    const targetInfo = {
      targetClassName: '',
      namespaceJoiner: ''
    }

    createRule(defaultMode, targetInfo)

    const styledComponent = React.forwardRef<HTMLElement, AnyObject>((props, ref) => {
      const { as = inputTag, className = '', variants: variantsProps, ...rest } = props
      let variantsClassName = ''

      const { mode } = React.useContext(themeContent)

      if (mode !== currentMode && mode !== undefined) {
        createRule(mode, targetInfo)
        currentMode = mode
      }

      if (variantsProps) {
        const variantsPropsKeys = Object.keys(variantsProps)
        let variantsPropsIndex = variantsPropsKeys.length

        while (variantsPropsIndex--) {
          const key = variantsPropsKeys[variantsPropsIndex]
          const value = variantsProps[key]

          variantsClassName += ` ${targetInfo.namespaceJoiner}${key}-${value}`
        }
      }

      return React.createElement(as, {
        className: `${className ? className + ' ' : className}${
          targetInfo.targetClassName
        }${variantsClassName}`,
        ref,
        ...rest
      })
    })

    if (process.env.NODE_ENV !== 'production') {
      styledComponent.displayName = styledComponent.displayName ?? targetInfo.targetClassName
    }

    Object.defineProperty(styledComponent, 'toString', {
      value() {
        return `.${targetInfo.targetClassName}`
      }
    })

    return styledComponent
  }

  function getCssValue() {
    return `
      <meta id=${cacheKey} name="styls-cache" content="${[...globalCache].join(',')}">
      <style data-styls="${sheet.key}">${sheet.ssrData}</style>
    `
  }

  function flush() {
    sheet.flush()
    globalCache.clear()
  }

  return { styled, SystemProvider, useSystem, getCssValue, flush }
}

export const { styled, getCssValue, flush } = createSystem()
