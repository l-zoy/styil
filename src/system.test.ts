import { createSystem, styled, flush, global } from './system'
import { render, fireEvent, getByText } from '@testing-library/react'
import React from 'react'

const meta = globalThis.document.createElement('meta')
const { document } = globalThis

jest.spyOn(globalThis.document, 'querySelector').mockImplementationOnce(() => {
  return meta
})

Object.defineProperty(globalThis, 'document', { value: undefined, writable: true })

describe('system', () => {
  beforeEach(() => {
    flush()
    globalThis.document = document
  })
  it('styled', () => {
    const Button = styled('button', {
      backgroundColor: 'gainsboro',
      borderRadius: '9999px',
      fontWeight: 500,
      padding: '0.75em 1em',
      border: 0,
      transition: 'all 200ms ease',

      '&:hover': {
        transform: 'translateY(-2px)',
        boxShadow: '0 10px 25px rgba(0, 0, 0, .3)'
      }
    })

    const { container } = render(React.createElement(Button))

    expect(container).toMatchSnapshot()

    expect(document.documentElement).toMatchSnapshot()
  })

  it('styled with styled components', () => {
    const Button = styled('button', {
      backgroundColor: '#fff'
    })

    const Button2 = styled(Button, {
      color: 'red'
    })

    const { container } = render(React.createElement(Button2))
    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
  })

  it('global', () => {
    global({
      body: {
        backgroundColor: '#fff'
      }
    })

    expect(document.documentElement).toMatchSnapshot()
  })

  it('with theme global', () => {
    const {
      global: globalTheme,
      useSystem,
      SystemProvider,
      flush
    } = createSystem({
      theme: (mode) => {
        return {
          dark: {
            color: 'red'
          },
          light: {
            color: 'blue'
          }
        }[mode]
      },
      defaultMode: 'light'
    })

    globalTheme((theme) => ({
      body: {
        backgroundColor: theme.color
      }
    }))

    const { container } = render(
      React.createElement(SystemProvider, {
        children: React.createElement(() => {
          const { mode, setMode } = useSystem()

          return React.createElement(
            'div',
            { onClick: () => setMode(mode === 'light' ? 'dark' : 'light') },
            mode
          )
        })
      })
    )

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    fireEvent.click(getByText(container, 'light'))
    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    fireEvent.click(getByText(container, 'dark'))
    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    flush()
  })

  it('delete old rules', () => {
    const {
      global: globalTheme,
      useSystem,
      SystemProvider,
      flush
    } = createSystem({
      theme: (mode) => {
        return {
          dark: {
            color: 'red'
          },
          light: {
            color: 'blue'
          }
        }[mode]
      },
      defaultMode: 'light',
      sheetOptions: { speedy: true }
    })

    globalTheme((theme) => ({
      body: {
        backgroundColor: theme.color
      }
    }))

    const { container } = render(
      React.createElement(SystemProvider, {
        children: React.createElement(() => {
          const { mode, setMode } = useSystem()

          return React.createElement(
            'div',
            { onClick: () => setMode(mode === 'light' ? 'dark' : 'light') },
            mode
          )
        })
      })
    )

    let styleHtml: HTMLStyleElement

    document.head.childNodes.forEach((item: HTMLStyleElement) => {
      styleHtml = item
    })

    expect(styleHtml.sheet.cssRules.length).toEqual(1)
    expect(styleHtml.sheet.cssRules[0].cssText).toEqual('body {background-color: blue;}')

    fireEvent.click(getByText(container, 'light'))
    expect(styleHtml.sheet.cssRules.length).toEqual(1)
    expect(styleHtml.sheet.cssRules[0].cssText).toEqual('body {background-color: red;}')
    fireEvent.click(getByText(container, 'dark'))
    expect(styleHtml.sheet.cssRules.length).toEqual(1)
    expect(styleHtml.sheet.cssRules[0].cssText).toEqual('body {background-color: blue;}')

    flush()
  })

  it('styled speedy', () => {
    const { flush, styled: thmeStyle } = createSystem({ sheetOptions: { speedy: true } })

    const Button = thmeStyle('button', {
      backgroundColor: 'gainsboro'
    })

    render(React.createElement(Button))

    document.head.childNodes.forEach((item: HTMLStyleElement) => {
      expect(item.sheet.cssRules.length).toEqual(1)
    })

    flush()
  })

  it('styled className', () => {
    const Button = styled('button', {
      backgroundColor: 'gainsboro'
    })

    const { container } = render(React.createElement(Button, { className: 'foo' }))

    expect(container).toMatchSnapshot()

    expect(document.documentElement).toMatchSnapshot()
  })

  it('styled toString', () => {
    const Button = styled('button', {
      backgroundColor: 'gainsboro',
      borderRadius: '9999px',
      fontWeight: 500,
      padding: '0.75em 1em',
      border: 0,
      transition: 'all 200ms ease',

      '&:hover': {
        transform: 'translateY(-2px)',
        boxShadow: '0 10px 25px rgba(0, 0, 0, .3)'
      }
    })

    expect(`${Button}`).toEqual('.css-3397010960')
  })

  it('styled production', () => {
    process.env.NODE_ENV = 'production'
    const Button = styled(
      'button',
      {
        borderRadius: '9999px'
      },
      {
        size: {
          small: {
            fontSize: 14
          }
        }
      }
    )

    const { container } = render(React.createElement(Button))

    expect(`${Button}`).toEqual('.css-3465392085')
    expect(container).toMatchSnapshot()

    process.env.NODE_ENV = 'test'
  })

  it('styled namespace', async () => {
    const { styled: _styled } = await import('./system')

    const Button = _styled(
      { tag: 'button', namespace: 'button' },
      {
        borderRadius: '9999px'
      }
    )

    const { container } = render(React.createElement(Button))

    expect(`${Button}`).toEqual('.button-css-3465392085')
    expect(container).toMatchSnapshot()

    process.env.NODE_ENV = 'test'
  })

  it('createSystem', async () => {
    const {
      styled: themeStyled,
      SystemProvider,
      useSystem,
      flush
    } = createSystem({
      theme: (mode) => {
        return {
          color: mode === 'light' ? 'red' : 'blue'
        }
      },
      defaultMode: 'light'
    })

    const Button = themeStyled('button', (theme) => ({
      color: theme.color
    }))

    const typemap = {
      dark: 'blue',
      light: 'red'
    }

    function Wapper() {
      const { mode, setMode, theme } = useSystem()

      expect(['light', 'dark'].includes(mode)).toEqual(true)
      expect(theme.color).toEqual(typemap[mode])

      return React.createElement(
        'div',
        { 'data-testid': 'use-system', onClick: () => setMode('dark') },
        React.createElement(Button),
        mode
      )
    }

    const { container } = render(
      React.createElement(SystemProvider, { children: React.createElement(Wapper) })
    )

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    fireEvent.click(getByText(container, 'light'))
    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    flush()
  })

  it('variants', async () => {
    const Button = styled(
      'button',
      {
        color: 'red'
      },
      {
        size: {
          small: {
            fontSize: 14
          },
          max: {
            fontSize: 99
          }
        }
      }
    )

    function Wapper() {
      const [state, setState] = React.useState<'small' | 'max'>('small')

      return React.createElement(
        'div',
        { 'data-testid': 'styled-variants' },
        React.createElement(
          Button,
          { variants: { size: state }, onClick: () => setState('max') },
          state
        )
      )
    }

    const { container } = render(React.createElement(Wapper))

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    fireEvent.click(getByText(container, 'small'))

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
  })

  it('variants single', async () => {
    const Button = styled(
      'button',
      {
        color: 'red'
      },
      {
        size: {
          small: {
            fontSize: 14
          }
        }
      }
    )

    function Wapper() {
      const [state, setState] = React.useState<'small' | undefined>('small')

      return React.createElement(
        'div',
        { 'data-testid': 'styled-variants' },
        React.createElement(
          Button,
          { variants: { size: state }, onClick: () => setState(undefined) },
          state
        )
      )
    }

    const { container } = render(React.createElement(Wapper))

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    fireEvent.click(getByText(container, 'small'))

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
  })

  it('variants theme', async () => {
    const {
      styled: themeStyled,
      SystemProvider,
      useSystem,
      flush
    } = createSystem({
      theme: (mode) => {
        return {
          color: mode === 'light' ? 'red' : 'blue'
        }
      },
      defaultMode: 'light'
    })

    const Button = themeStyled(
      'button',
      {
        color: 'red'
      },
      (theme) => ({
        color: {
          dark: {
            color: theme.color
          },
          light: {
            color: theme.color
          }
        }
      })
    )

    const typemap = {
      dark: 'blue',
      light: 'red'
    }

    function Wapper() {
      const { mode, setMode, theme } = useSystem()

      expect(['light', 'dark'].includes(mode)).toEqual(true)
      expect(theme.color).toEqual(typemap[mode])

      return React.createElement(
        'div',
        { 'data-testid': 'use-system', onClick: () => setMode('dark') },
        React.createElement(Button, { variants: { color: mode as 'light' } }),
        mode
      )
    }

    const { container } = render(
      React.createElement(SystemProvider, { children: React.createElement(Wapper) })
    )

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    fireEvent.click(getByText(container, 'light'))
    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    flush()
  })

  it('theme does not render repeatedly', async () => {
    const {
      styled: themeStyled,
      SystemProvider,
      useSystem,
      flush
    } = createSystem({
      theme: (mode) => {
        return {
          color: mode === 'light' ? 'red' : 'blue'
        }
      },
      defaultMode: 'light'
    })

    const Button = themeStyled('button', (theme) => ({
      color: theme.color
    }))

    const typemap = {
      dark: 'blue',
      light: 'red'
    }

    function Wapper() {
      const { mode, setMode, theme } = useSystem()

      expect(['light', 'dark'].includes(mode)).toEqual(true)
      expect(theme.color).toEqual(typemap[mode])

      return React.createElement(
        'div',
        { 'data-testid': 'use-system', onClick: () => setMode(mode === 'dark' ? 'light' : 'dark') },
        React.createElement(Button),
        mode
      )
    }

    const { container } = render(
      React.createElement(SystemProvider, { children: React.createElement(Wapper) })
    )

    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    fireEvent.click(getByText(container, 'light'))
    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    fireEvent.click(getByText(container, 'dark'))
    expect(container).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    flush()
  })
})

describe('hydrate', () => {
  beforeEach(() => {
    globalThis.document = undefined
  })
  it('base hydrate', async () => {
    meta.name = 'styils-cache'
    meta.content = `css-2242710476`
    document.head.appendChild(meta)

    const { styled: hydrateStyled, getCssValue } = createSystem()
    hydrateStyled('button', {
      backgroundColor: 'gainsboro',
      borderRadius: '9999px'
    })

    expect(getCssValue().html).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    document.head.removeChild(meta)
  })

  it('variants hydrate', () => {
    meta.name = 'styils-cache'
    meta.content = `css-2242710476|css-2242710476.size-max|css-2242710476.size-small`
    document.head.appendChild(meta)

    const { styled: hydrateStyled, getCssValue } = createSystem()
    hydrateStyled(
      'button',
      {
        backgroundColor: 'gainsboro',
        borderRadius: '9999px'
      },
      {
        size: {
          small: {
            fontSize: 14
          },
          max: {
            fontSize: 12
          }
        }
      }
    )

    expect(getCssValue().html).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    document.head.removeChild(meta)
  })

  it('namespace hydrate', () => {
    meta.name = 'styils-cache'
    meta.content = `ssr-css-2242710476|ssr-css-2242710476.ssr-size-max|ssr-css-2242710476.ssr-size-small`
    document.head.appendChild(meta)

    const { styled: hydrateStyled, getCssValue } = createSystem()
    hydrateStyled(
      { tag: 'button', namespace: 'ssr' },
      {
        backgroundColor: 'gainsboro',
        borderRadius: '9999px'
      },
      {
        size: {
          small: {
            fontSize: 14
          },
          max: {
            fontSize: 12
          }
        }
      }
    )

    expect(getCssValue().html).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    document.head.removeChild(meta)
  })

  it('global hydrate', () => {
    meta.name = 'styils-cache'
    meta.content = ''
    meta.setAttribute('mode', 'none')
    document.head.appendChild(meta)

    const { global: hydrateGlobal, getCssValue } = createSystem()
    hydrateGlobal({
      body: {
        backgroundColor: 'gainsboro',
        borderRadius: '9999px'
      }
    })

    expect(getCssValue().html).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()

    globalThis.document = document
    const { global: hydrateGlobal1, getCssValue: getCssValue1 } = createSystem()
    hydrateGlobal1({
      body: {
        backgroundColor: 'red',
        borderRadius: '99px'
      }
    })

    expect(getCssValue1().html).toMatchSnapshot()
    expect(document.documentElement).toMatchSnapshot()
    document.head.removeChild(meta)
  })

  it('render Element hydrate', async () => {
    const { styled: hydrateStyled, getCssValue } = createSystem()

    hydrateStyled('div', {
      backgroundColor: 'gainsboro',
      borderRadius: '9999px'
    })
    const { StyilRules } = getCssValue()

    globalThis.document = document
    const { container } = render(StyilRules)

    expect(container).toMatchSnapshot()
  })

  it('mix hydrate', async () => {
    globalThis.document = document

    styled('div', {
      backgroundColor: 'gainsboro',
      borderRadius: '9999px'
    })
    Object.defineProperty(globalThis, 'document', { value: undefined })
    const { styled: hydrateStyled, getCssValue } = createSystem()

    hydrateStyled('div', {
      backgroundColor: 'gainsboro',
      borderRadius: '9999px'
    })

    expect(getCssValue().html).toMatchSnapshot()
  })
})
