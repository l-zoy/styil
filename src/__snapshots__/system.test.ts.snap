// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`system createSystem 1`] = `
<div>
  <div
    data-testid="use-system"
  >
    <button
      class="css-3433634237"
    />
    light
  </div>
</div>
`;

exports[`system createSystem 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="use-system"
      >
        <button
          class="css-3433634237"
        />
        light
      </div>
    </div>
  </body>
</html>
`;

exports[`system createSystem 3`] = `
<div>
  <div
    data-testid="use-system"
  >
    <button
      class="css-3183262666"
    />
    dark
  </div>
</div>
`;

exports[`system createSystem 4`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3183262666{color:blue;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="use-system"
      >
        <button
          class="css-3183262666"
        />
        dark
      </div>
    </div>
  </body>
</html>
`;

exports[`system global 1`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      body{background-color:#fff;}
    </style>
  </head>
  <body />
</html>
`;

exports[`system global with theme 1`] = `
<div>
  <div>
    light
  </div>
</div>
`;

exports[`system global with theme 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      body{background-color:blue;}
    </style>
  </head>
  <body>
    <div>
      <div>
        light
      </div>
    </div>
  </body>
</html>
`;

exports[`system global with theme 3`] = `
<div>
  <div>
    dark
  </div>
</div>
`;

exports[`system global with theme 4`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      body{background-color:blue;}
    </style>
    <style
      data-styil="css"
    >
      
      body{background-color:red;}
    </style>
  </head>
  <body>
    <div>
      <div>
        dark
      </div>
    </div>
  </body>
</html>
`;

exports[`system hydrate 1`] = `
"
      <meta id=\\"__styil_cache__\\" name=\\"styil-cache\\" content=\\"css-2242710476\\">
      <style data-styil=\\"css\\"></style>
    "
`;

exports[`system hydrate 2`] = `
<html>
  <head>
    <meta
      content="css-2242710476"
      id="__styil_cache__"
      name="styil-cache"
    />
  </head>
  <body>
    <div>
      <button
        class="css-2242710476"
      />
    </div>
  </body>
</html>
`;

exports[`system hydrate namespace 1`] = `
"
      <meta id=\\"__styil_cache__\\" name=\\"styil-cache\\" content=\\"ssr-css-2242710476|ssr-css-2242710476.ssr-size-max|ssr-css-2242710476.ssr-size-small\\">
      <style data-styil=\\"css\\"></style>
    "
`;

exports[`system hydrate namespace 2`] = `
<html>
  <head>
    <meta
      content="ssr-css-2242710476|ssr-css-2242710476.ssr-size-max|ssr-css-2242710476.ssr-size-small"
      id="__styil_cache__"
      name="styil-cache"
    />
  </head>
  <body>
    <div>
      <button
        class="ssr-css-2242710476 ssr-size-small"
      />
    </div>
  </body>
</html>
`;

exports[`system hydrate variants 1`] = `
"
      <meta id=\\"__styil_cache__\\" name=\\"styil-cache\\" content=\\"css-2242710476|css-2242710476.size-max|css-2242710476.size-small\\">
      <style data-styil=\\"css\\"></style>
    "
`;

exports[`system hydrate variants 2`] = `
<html>
  <head>
    <meta
      content="css-2242710476|css-2242710476.size-max|css-2242710476.size-small"
      id="__styil_cache__"
      name="styil-cache"
    />
  </head>
  <body>
    <div>
      <button
        class="css-2242710476"
      />
    </div>
  </body>
</html>
`;

exports[`system styled 1`] = `
<div>
  <button
    class="css-3397010960"
  />
</div>
`;

exports[`system styled 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3397010960{background-color:gainsboro;border-radius:9999px;font-weight:500;padding:0.75em 1em;border:0px;transition:all 200ms ease;}.css-3397010960:hover{transform:translateY(-2px);box-shadow:0 10px 25px rgba(0, 0, 0, .3);}
    </style>
  </head>
  <body>
    <div>
      <button
        class="css-3397010960"
      />
    </div>
  </body>
</html>
`;

exports[`system styled className 1`] = `
<div>
  <button
    class="foo css-2129283522"
  />
</div>
`;

exports[`system styled className 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-2129283522{background-color:gainsboro;}
    </style>
  </head>
  <body>
    <div>
      <button
        class="foo css-2129283522"
      />
    </div>
  </body>
</html>
`;

exports[`system styled namespace 1`] = `
<div>
  <button
    class="button-css-3465392085"
  />
</div>
`;

exports[`system styled production 1`] = `
<div>
  <button
    class="css-3465392085"
  />
</div>
`;

exports[`system styled with styled components 1`] = `
<div>
  <button
    class="css-3433634237 css-2341007075"
  />
</div>
`;

exports[`system styled with styled components 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-2341007075{background-color:#fff;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
  </head>
  <body>
    <div>
      <button
        class="css-3433634237 css-2341007075"
      />
    </div>
  </body>
</html>
`;

exports[`system theme does not render repeatedly 1`] = `
<div>
  <div
    data-testid="use-system"
  >
    <button
      class="css-3433634237"
    />
    light
  </div>
</div>
`;

exports[`system theme does not render repeatedly 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="use-system"
      >
        <button
          class="css-3433634237"
        />
        light
      </div>
    </div>
  </body>
</html>
`;

exports[`system theme does not render repeatedly 3`] = `
<div>
  <div
    data-testid="use-system"
  >
    <button
      class="css-3183262666"
    />
    dark
  </div>
</div>
`;

exports[`system theme does not render repeatedly 4`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3183262666{color:blue;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="use-system"
      >
        <button
          class="css-3183262666"
        />
        dark
      </div>
    </div>
  </body>
</html>
`;

exports[`system theme does not render repeatedly 5`] = `
<div>
  <div
    data-testid="use-system"
  >
    <button
      class="css-3433634237"
    />
    light
  </div>
</div>
`;

exports[`system theme does not render repeatedly 6`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3183262666{color:blue;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="use-system"
      >
        <button
          class="css-3433634237"
        />
        light
      </div>
    </div>
  </body>
</html>
`;

exports[`system variants 1`] = `
<div>
  <div
    data-testid="styled-variants"
  >
    <button
      class="css-3433634237 size-small"
    >
      small
    </button>
  </div>
</div>
`;

exports[`system variants 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.size-max{font-size:99px;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.size-small{font-size:14px;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="styled-variants"
      >
        <button
          class="css-3433634237 size-small"
        >
          small
        </button>
      </div>
    </div>
  </body>
</html>
`;

exports[`system variants 3`] = `
<div>
  <div
    data-testid="styled-variants"
  >
    <button
      class="css-3433634237 size-max"
    >
      max
    </button>
  </div>
</div>
`;

exports[`system variants 4`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.size-max{font-size:99px;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.size-small{font-size:14px;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="styled-variants"
      >
        <button
          class="css-3433634237 size-max"
        >
          max
        </button>
      </div>
    </div>
  </body>
</html>
`;

exports[`system variants single 1`] = `
<div>
  <div
    data-testid="styled-variants"
  >
    <button
      class="css-3433634237 size-small"
    >
      small
    </button>
  </div>
</div>
`;

exports[`system variants single 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.size-small{font-size:14px;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="styled-variants"
      >
        <button
          class="css-3433634237 size-small"
        >
          small
        </button>
      </div>
    </div>
  </body>
</html>
`;

exports[`system variants single 3`] = `
<div>
  <div
    data-testid="styled-variants"
  >
    <button
      class="css-3433634237"
    />
  </div>
</div>
`;

exports[`system variants single 4`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.size-small{font-size:14px;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="styled-variants"
      >
        <button
          class="css-3433634237"
        />
      </div>
    </div>
  </body>
</html>
`;

exports[`system variants theme 1`] = `
<div>
  <div
    data-testid="use-system"
  >
    <button
      class="css-3433634237 color-light"
    />
    light
  </div>
</div>
`;

exports[`system variants theme 2`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.color-light{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.color-dark{color:red;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="use-system"
      >
        <button
          class="css-3433634237 color-light"
        />
        light
      </div>
    </div>
  </body>
</html>
`;

exports[`system variants theme 3`] = `
<div>
  <div
    data-testid="use-system"
  >
    <button
      class="css-3433634237 color-dark"
    />
    dark
  </div>
</div>
`;

exports[`system variants theme 4`] = `
<html>
  <head>
    <style
      data-styil="css"
    >
      
      .css-3433634237{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.color-light{color:red;}
    </style>
    <style
      data-styil="css"
    >
      
      .css-3433634237.color-dark{color:red;}
    </style>
  </head>
  <body>
    <div>
      <div
        data-testid="use-system"
      >
        <button
          class="css-3433634237 color-dark"
        />
        dark
      </div>
    </div>
  </body>
</html>
`;
